File: src/buildings/queen.js
import * as PIXI from 'pixi.js'
import { Building, keyToCoords, toKey } from '../building.js'

export class Queen extends Building {

    constructor(game) {
        super('Queen',{x:3, y:3},[[1,1,1],[1,0,1],[1,1,1]],game,true)
        this.sprite = PIXI.Sprite.from('Queen')

        this.algaeQuota = 20
        this.algaeCount = 0
        this.broodlingCount = 1

        this.description = `The one and only Queen of your colony! Protect her at all costs!`
    }

    getAlgaeQuota() {
        return this.algaeQuota
    }

    getAlgaeCount() {
        return this.algaeCount
    }

    getFeedTiles() {
        const feedTiles = []
        for (const tile of this.tileArray) {
            if (tile && tile.creatureFits()) {
                feedTiles.push(tile)
            }
        }
        return feedTiles
    }

    canBeFedBy(creature) {
        if (!creature || !creature.location) {
            return false
        }

        const creatureKey = toKey(creature.location)
        for (const tile of this.getFeedTiles()) {
            if (tile.key === creatureKey) {
                return true
            }
        }

        return false
    }

    getBirthTile() {
        const feedTiles = this.getFeedTiles()
        if (feedTiles.length === 0) {
            return null
        }

        const randomIndex = Math.floor(Math.random() * feedTiles.length)
        return feedTiles[randomIndex]
    }

    birth(cave, feeder = null) {
        const activeCave = cave ?? feeder?.cave
        if (!activeCave || !feeder || typeof feeder.constructor !== 'function') {
            return false
        }

        const birthTile = this.getBirthTile()
        if (!birthTile || !birthTile.creatureFits()) {
            return false
        }

        const spawnLocation = keyToCoords(birthTile.key)
        const spawnName = `Broodling ${this.broodlingCount}`
        this.broodlingCount++

        const brood = new feeder.constructor(spawnName, spawnLocation, feeder.game)
        return activeCave.spawn(brood, birthTile)
    }

    feedAlgae(amount, creature = null, cave = null) {
        if (!Number.isFinite(amount) || amount <= 0) {
            return { accepted: 0, spawnCount: 0 }
        }

        if (creature && !this.canBeFedBy(creature)) {
            return { accepted: 0, spawnCount: 0 }
        }

        this.algaeCount += amount

        let spawnCount = 0
        while (this.algaeCount >= this.algaeQuota) {
            this.algaeCount -= this.algaeQuota
            this.algaeQuota += 5
            if (this.birth(cave, creature)) {
                spawnCount++
            }
        }

        return {
            accepted: amount,
            spawnCount
        }
    }
}

File: TriloGame.CSharp/src/TriloGame.Game/Core/Buildings/Queen.cs
using TriloGame.Game.Audio;
using TriloGame.Game.Core.Entities;
using TriloGame.Game.Core.Simulation;
using TriloGame.Game.Shared.Math;
using TriloGame.Game.Shared.Utilities;

namespace TriloGame.Game.Core.Buildings;

public sealed class Queen : Building
{
    public Queen(GameSession session)
        : base("Queen", new GridPoint(3, 3), [[1, 1, 1], [1, 0, 1], [1, 1, 1]], session, true)
    {
        TextureKey = "Queen";
        AlgaeQuota = 20;
        AlgaeCount = 0;
        BroodlingCount = 1;
        Description = "The one and only Queen of your colony! Protect her at all costs!";
    }

    public int AlgaeQuota { get; private set; }

    public int AlgaeCount { get; private set; }

    public int BroodlingCount { get; private set; }

    public IReadOnlyList<World.Tile> GetFeedTiles()
    {
        return TileArray.Where(tile => tile.CreatureFits()).ToArray();
    }

    public bool CanBeFedBy(Creature creature)
    {
        return GetFeedTiles().Any(tile => tile.Key == creature.Location.ToString());
    }

    public World.Tile? GetBirthTile()
    {
        var feedTiles = GetFeedTiles();
        return feedTiles.Count == 0 ? null : feedTiles[RandomUtil.NextInt(feedTiles.Count)];
    }

    public bool Birth(World.Cave? cave, Trilobite? feeder)
    {
        if (cave is null || feeder is null)
        {
            return false;
        }

        var birthTile = GetBirthTile();
        if (birthTile is null || !birthTile.CreatureFits())
        {
            return false;
        }

        var spawnName = $"Broodling {BroodlingCount}";
        BroodlingCount++;
        var brood = new Trilobite(spawnName, GridPoint.Parse(birthTile.Key), feeder.Session);
        var spawned = cave.Spawn(brood, birthTile);
        if (spawned)
        {
            Session.RequestAudioCue(GameAudioCue.TrilobiteBirth);
        }

        return spawned;
    }

    public (int Accepted, int SpawnCount) FeedAlgae(int amount, Trilobite? creature = null, World.Cave? cave = null)
    {
        if (amount <= 0)
        {
            return (0, 0);
        }

        if (creature is not null && !CanBeFedBy(creature))
        {
            return (0, 0);
        }

        AlgaeCount += amount;
        var spawnCount = 0;
        while (AlgaeCount >= AlgaeQuota)
        {
            AlgaeCount -= AlgaeQuota;
            AlgaeQuota += 5;
            if (Birth(cave, creature))
            {
                spawnCount++;
            }
        }

        return (amount, spawnCount);
    }
}
